core-bundle/src/Resources/contao/controllers/FrontendIndex.php

```diff
@@ -117,27 +62,6 @@ public function renderPage(PageModel $pageModel)
 			list($GLOBALS['TL_ADMIN_NAME'], $GLOBALS['TL_ADMIN_EMAIL']) = StringUtil::splitFriendlyEmail(Config::get('adminEmail'));
 		}
 
-		// Exit if the root page has not been published (see #2425)
-		// Do not try to load the 404 page, it can cause an infinite loop!
-		if (!$blnShowUnpublished && !$objPage->rootIsPublic)
-		{
-			throw new PageNotFoundException('Page not found: ' . Environment::get('uri'));
-		}
-
-		// Check whether the language matches the root page language
-		if (isset($_GET['language']) && $objPage->urlPrefix && Input::get('language') != LocaleUtil::formatAsLanguageTag($objPage->rootLanguage))
-		{
-			throw new PageNotFoundException('Page not found: ' . Environment::get('uri'));
-		}
-
-		// Check whether there are domain name restrictions
-		if ($objPage->domain && $objPage->domain != Environment::get('host'))
-		{
-			System::getContainer()->get('monolog.logger.contao.error')->error('Page ID "' . $objPage->id . '" was requested via "' . Environment::get('host') . '" but can only be accessed via "' . $objPage->domain . '" (' . Environment::get('base') . Environment::get('request') . ')');
-
-			throw new PageNotFoundException('Page not found: ' . Environment::get('uri'));
-		}
-
```

Comment:
Are you 100% sure that all this (and the lines above) can be removed? All the checks look valid to me.

Reply:
they are all done by the routing component. A page is never matched if it is not published, the prefix does not match or the domain is different.